contracts/governor/src/contract.rs

```diff
@@ -1,4 +1,209 @@
+        for call_data in proposal.sub_calldata {
+            let pre_auth_entry = InvokerContractAuthEntry::Contract(SubContractInvocation {
+                context: ContractContext {
+                    contract: call_data.contract_id,
+                    fn_name: call_data.function,
+                    args: call_data.args,
+                },
+                sub_invocations: vec![&e],
+            });
+            pre_auth_vec.push_back(pre_auth_entry);
+        }
+        e.authorize_as_current_contract(pre_auth_vec);
+        let _: () = e.invoke_contract(
+            &proposal.calldata.contract_id,
+            &proposal.calldata.function,
+            proposal.calldata.args,
+        );
+        storage::set_proposal_status(&e, &proposal_id, &ProposalStatus::Succeeded);
+    }
+
+    fn cancel(e: Env, creator: Address, proposal_id: u32) {
+        creator.require_auth();
+        let proposal = storage::get_proposal(&e, &proposal_id);
+        if proposal.is_none() {
+            panic_with_error!(&e, GovernorError::NonExistentProposalError);
+        }
+        let proposal_status = storage::get_proposal_status(&e, &proposal_id);
+        if proposal_status != ProposalStatus::Pending {
+            panic_with_error!(&e, GovernorError::CancelActiveProposalError);
+        }
+        storage::set_proposal_status(&e, &proposal_id, &ProposalStatus::Expired);
+    }
+
+    fn vote(e: Env, voter: Address, proposal_id: u32, support: u32) {
+        voter.require_auth();
+        let proposal = storage::get_proposal(&e, &proposal_id);
+        if proposal.is_none() {
+            panic_with_error!(&e, GovernorError::NonExistentProposalError);
+        }
+
+        let proposal = proposal.unwrap_optimized();
+        let proposal_status = storage::get_proposal_status(&e, &proposal_id);
+        if proposal_status != ProposalStatus::Active {
+            if proposal_status == ProposalStatus::Pending
+                && proposal.vote_start <= e.ledger().timestamp()
+            {
+                storage::set_proposal_status(&e, &proposal_id, &ProposalStatus::Active)
+            } else {
+                panic_with_error!(&e, GovernorError::ProposalNotActiveError);
+            }
+        }
+
+        let voter_power = VotesClient::new(&e, &storage::get_voter_token_address(&e))
+            .get_past_votes(&voter, &proposal.vote_start);
+        let mut vote_count = storage::get_proposal_vote_count(&e, &proposal_id);
+        let voter_status = storage::get_voter_status(&e, &voter, &proposal_id);
+
+        // Check if voter has already voted and remove previous vote from count
+        if let Some(voter_status) = voter_status {
+            match voter_status {
+                0 => vote_count.votes_abstained -= voter_power,
+                1 => vote_count.votes_against -= voter_power,
+                2 => vote_count.votes_for -= voter_power,
+                _ => (),
+            }
+        }
```

Comment:
can we make an issue for this? I'm not sure we want people to be able to change their vote after submission.

Would be worth reviewing other DAO implementations to see if this is common (offhand I don't think it is)